In [1]:
#Para manejos de directorios y archivos
import os

# Computacion numerica en Python, como arrays y matrices
import numpy as np

#Proporciona estrucutra de datos y herramientas para el analisis de datos
import pandas as pd

#Datos en una tabla
from tabulate import tabulate

#Para graficación, visulaizacion de datos
from matplotlib import pyplot
#Superficies en 3D
from mpl_toolkits.mplot3d import Axes3D

#Para separar el 20% y 80%
from sklearn.model_selection import train_test_split

#Modulo de optimizacion en scipy
from scipy import optimize
from scipy.io import loadmat

#Visualizar graficas en linea
%matplotlib inline

In [2]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


Red Neuronal

In [3]:
dataset = pd.read_csv('/content/gdrive/MyDrive/SIS 420 INTELIGENCIA ARTIFICAL/datasets/covtype.csv', delimiter=',')
dataset

Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0            2596      51      3                               258   
1            2590      56      2                               212   
2            2804     139      9                               268   
3            2785     155     18                               242   
4            2595      45      2                               153   
...           ...     ...    ...                               ...   
581007       2396     153     20                                85   
581008       2391     152     19                                67   
581009       2386     159     17                                60   
581010       2384     170     15                                60   
581011       2383     165     13                                60   

        Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                                    0                              510   
1                                   -6                              390   
2                                   65                             3180   
3                                  118                             3090   
4                                   -1                              391   
...                                ...                              ...   
581007                              17                              108   
581008                              12                               95   
581009                               7                               90   
581010                               5                               90   
581011                               4                               67   

        Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
0                 221             232            148   
1                 220             235            151   
2                 234             238            135   
3                 238             238            122   
4                 220             234            150   
...               ...             ...            ...   
581007            240             237            118   
581008            240             237            119   
581009            236             241            130   
581010            230             245            143   
581011            231             244            141   

        Horizontal_Distance_To_Fire_Points  ...  Soil_Type32  Soil_Type33  \
0                                     6279  ...            0            0   
1                                     6225  ...            0            0   
2                                     6121  ...            0            0   
3                                     6211  ...            0            0   
4                                     6172  ...            0            0   
...                                    ...  ...          ...          ...   
581007                                 837  ...            0            0   
581008                                 845  ...            0            0   
581009                                 854  ...            0            0   
581010                                 864  ...            0            0   
581011                                 875  ...            0            0   

        Soil_Type34  Soil_Type35  Soil_Type36  Soil_Type37  Soil_Type38  \
0                 0            0            0            0            0   
1                 0            0            0            0            0   
2                 0            0            0            0            0   
3                 0            0            0            0            0   
4                 0            0            0            0            0   
...             ...          ...          ...          ...          ...   
581007            0            0            0            0            0   
581008            0            0            0            0          

Analisis del Dataset

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 581012 entries, 0 to 581011
Data columns (total 55 columns):
 #   Column                              Non-Null Count   Dtype
---  ------                              --------------   -----
 0   Elevation                           581012 non-null  int64
 1   Aspect                              581012 non-null  int64
 2   Slope                               581012 non-null  int64
 3   Horizontal_Distance_To_Hydrology    581012 non-null  int64
 4   Vertical_Distance_To_Hydrology      581012 non-null  int64
 5   Horizontal_Distance_To_Roadways     581012 non-null  int64
 6   Hillshade_9am                       581012 non-null  int64
 7   Hillshade_Noon                      581012 non-null  int64
 8   Hillshade_3pm                       581012 non-null  int64
 9   Horizontal_Distance_To_Fire_Points  581012 non-null  int64
 10  Wilderness_Area1                    581012 non-null  int64
 11  Wilderness_Area2                    581012 non-null 

In [5]:
#para contar cuantas clases contiene nuestra columna y
class_counts = dataset["Cover_Type"].value_counts()

#para contar cuantas caracteristicas tiene nuestro dataset, obviamente con sin contar nuestra y, por eso lo dropeamos, tambien dropeamos la primera columna que no tiene nombre
feactures_counts =dataset.drop(['Cover_Type'], axis=1)
feactures_counts = feactures_counts.shape[1]

print(f"la cantidad de caracteristicas es: { feactures_counts}")
print(f'la cantidad de clases es: 8')

#mostramos la cantidad de clases tiene, y en que cantidad
class_counts

la cantidad de caracteristicas es: 54
la cantidad de clases es: 8


Cover_Type
2    283301
1    211840
3     35754
7     20510
6     17367
5      9493
4      2747
Name: count, dtype: int64

In [6]:
dataset['Cover_Type'] = dataset['Cover_Type'].replace(7, 0)

In [7]:
dataset

Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0            2596      51      3                               258   
1            2590      56      2                               212   
2            2804     139      9                               268   
3            2785     155     18                               242   
4            2595      45      2                               153   
...           ...     ...    ...                               ...   
581007       2396     153     20                                85   
581008       2391     152     19                                67   
581009       2386     159     17                                60   
581010       2384     170     15                                60   
581011       2383     165     13                                60   

        Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                                    0                              510   
1                                   -6                              390   
2                                   65                             3180   
3                                  118                             3090   
4                                   -1                              391   
...                                ...                              ...   
581007                              17                              108   
581008                              12                               95   
581009                               7                               90   
581010                               5                               90   
581011                               4                               67   

        Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
0                 221             232            148   
1                 220             235            151   
2                 234             238            135   
3                 238             238            122   
4                 220             234            150   
...               ...             ...            ...   
581007            240             237            118   
581008            240             237            119   
581009            236             241            130   
581010            230             245            143   
581011            231             244            141   

        Horizontal_Distance_To_Fire_Points  ...  Soil_Type32  Soil_Type33  \
0                                     6279  ...            0            0   
1                                     6225  ...            0            0   
2                                     6121  ...            0            0   
3                                     6211  ...            0            0   
4                                     6172  ...            0            0   
...                                    ...  ...          ...          ...   
581007                                 837  ...            0            0   
581008                                 845  ...            0            0   
581009                                 854  ...            0            0   
581010                                 864  ...            0            0   
581011                                 875  ...            0            0   

        Soil_Type34  Soil_Type35  Soil_Type36  Soil_Type37  Soil_Type38  \
0                 0            0            0            0            0   
1                 0            0            0            0            0   
2                 0            0            0            0            0   
3                 0            0            0            0            0   
4                 0            0            0            0            0   
...             ...          ...          ...          ...          ...   
581007            0            0            0            0            0   
581008            0            0            0            0          

In [8]:
#para contar cuantas clases contiene nuestra columna y
class_counts = dataset["Cover_Type"].value_counts()

#para contar cuantas caracteristicas tiene nuestro dataset, obviamente con sin contar nuestra y, por eso lo dropeamos, tambien dropeamos la primera columna que no tiene nombre
feactures_counts =dataset.drop(['Cover_Type'], axis=1)
feactures_counts = feactures_counts.shape[1]

print(f"la cantidad de caracteristicas es: { feactures_counts}")
print(f'la cantidad de clases es: 8')

#mostramos la cantidad de clases tiene, y en que cantidad
class_counts

la cantidad de caracteristicas es: 54
la cantidad de clases es: 8


Cover_Type
2    283301
1    211840
3     35754
0     20510
6     17367
5      9493
4      2747
Name: count, dtype: int64

In [9]:
#almacenamos la cantidad de caracteristicas en una variable
input_layer_size  = feactures_counts;

#almacenamos la cantidad de clases en una variable

num_labels = 7;

In [10]:
# hacemos uso del DataFrame llamado 'dataset' que contiene nuestros datos datos
# y es la columna que contiene las etiquetas de clase

#creamos una variable temporal que contentra toda la columna de 'Credit_Score'
y_temp = dataset['Cover_Type']

# Para la clase 0

#donde y_temp es igual a 0, separamos los datos en train_class_0, test_class_0
data_class_0 = dataset[y_temp == 0]
train_class_0, test_class_0 = train_test_split(data_class_0, test_size=0.2, random_state=42)

# Para la clase 1

#donde y_temp es igual a 1, separamos los datos en train_class_1, test_class_1
data_class_1 = dataset[y_temp == 1]
train_class_1, test_class_1 = train_test_split(data_class_1, test_size=0.2, random_state=42)

# Para la clase 2

data_class_2 = dataset[y_temp == 2]
train_class_2, test_class_2 = train_test_split(data_class_2, test_size=0.2, random_state=42)

# Para la clase 3

data_class_3 = dataset[y_temp == 3]
train_class_3, test_class_3 = train_test_split(data_class_3, test_size=0.2, random_state=42)

# Para la clase 4

data_class_4 = dataset[y_temp == 4]
train_class_4, test_class_4 = train_test_split(data_class_4, test_size=0.2, random_state=42)

# Para la clase 5

data_class_5 = dataset[y_temp == 5]
train_class_5, test_class_5 = train_test_split(data_class_5, test_size=0.2, random_state=42)

# Para la clase 6

data_class_6 = dataset[y_temp == 6]
train_class_6, test_class_6 = train_test_split(data_class_6, test_size=0.2, random_state=42)

In [11]:
print(f"Para la clase 0 tenemos una cantidad de: { data_class_0.shape[0]} donde el 80% es: {train_class_0.shape[0]} y el 20% es: {test_class_0.shape[0]}")
print(f"Para la clase 1 tenemos una cantidad de: { data_class_1.shape[0]} donde el 80% es: {train_class_1.shape[0]} y el 20% es: {test_class_1.shape[0]}")
print(f"Para la clase 2 tenemos una cantidad de: { data_class_2.shape[0]} donde el 80% es: {train_class_2.shape[0]} y el 20% es: {test_class_2.shape[0]}")
print(f"Para la clase 3 tenemos una cantidad de: { data_class_3.shape[0]} donde el 80% es: {train_class_3.shape[0]} y el 20% es: {test_class_3.shape[0]}")
print(f"Para la clase 4 tenemos una cantidad de: { data_class_4.shape[0]} donde el 80% es: {train_class_4.shape[0]} y el 20% es: {test_class_4.shape[0]}")
print(f"Para la clase 5 tenemos una cantidad de: { data_class_5.shape[0]} donde el 80% es: {train_class_5.shape[0]} y el 20% es: {test_class_5.shape[0]}")
print(f"Para la clase 6 tenemos una cantidad de: { data_class_6.shape[0]} donde el 80% es: {train_class_6.shape[0]} y el 20% es: {test_class_6.shape[0]}")
print(f"La cantidad total de datos es: {dataset.shape[0]}")

Para la clase 0 tenemos una cantidad de: 20510 donde el 80% es: 16408 y el 20% es: 4102
Para la clase 1 tenemos una cantidad de: 211840 donde el 80% es: 169472 y el 20% es: 42368
Para la clase 2 tenemos una cantidad de: 283301 donde el 80% es: 226640 y el 20% es: 56661
Para la clase 3 tenemos una cantidad de: 35754 donde el 80% es: 28603 y el 20% es: 7151
Para la clase 4 tenemos una cantidad de: 2747 donde el 80% es: 2197 y el 20% es: 550
Para la clase 5 tenemos una cantidad de: 9493 donde el 80% es: 7594 y el 20% es: 1899
Para la clase 6 tenemos una cantidad de: 17367 donde el 80% es: 13893 y el 20% es: 3474
La cantidad total de datos es: 581012


In [12]:
#para la parte de entrenamiento, separamos las caracteristicas de la etiqueta

#para la clase 0
X_train_class_0 = train_class_0.drop(['Cover_Type'], axis=1)
y_train_class_0 = train_class_0['Cover_Type']

#para la clase 1
X_train_class_1 = train_class_1.drop(['Cover_Type'], axis=1)
y_train_class_1 = train_class_1['Cover_Type']

#para la clase 2
X_train_class_2 = train_class_2.drop(['Cover_Type'], axis=1)
y_train_class_2 = train_class_2['Cover_Type']

#para la clase 3
X_train_class_3 = train_class_3.drop(['Cover_Type'], axis=1)
y_train_class_3 = train_class_3['Cover_Type']

#para la clase 4
X_train_class_4 = train_class_4.drop(['Cover_Type'], axis=1)
y_train_class_4 = train_class_4['Cover_Type']

#para la clase 5
X_train_class_5 = train_class_5.drop(['Cover_Type'], axis=1)
y_train_class_5 = train_class_5['Cover_Type']

#para la clase 6
X_train_class_6 = train_class_6.drop(['Cover_Type'], axis=1)
y_train_class_6 = train_class_6['Cover_Type']

#ahora para la parte de pruebas, separamos las caracteristicas de la etiqueta
#para la clase 0
X_test_class_0 = test_class_0.drop(['Cover_Type'], axis=1)
y_test_class_0 = test_class_0['Cover_Type']

#para la clase 1
X_test_class_1 = test_class_1.drop(['Cover_Type'], axis=1)
y_test_class_1 = test_class_1['Cover_Type']

#para la clase 2
X_test_class_2 = test_class_2.drop(['Cover_Type'], axis=1)
y_test_class_2 = test_class_2['Cover_Type']

#para la clase 3
X_test_class_3 = test_class_3.drop(['Cover_Type'], axis=1)
y_test_class_3 = test_class_3['Cover_Type']

#para la clase 4
X_test_class_4 = test_class_4.drop(['Cover_Type'], axis=1)
y_test_class_4 = test_class_4['Cover_Type']

#para la clase 5
X_test_class_5 = test_class_5.drop(['Cover_Type'], axis=1)
y_test_class_5 = test_class_5['Cover_Type']

#para la clase 6
X_test_class_6 = test_class_6.drop(['Cover_Type'], axis=1)
y_test_class_6 = test_class_6['Cover_Type']

In [13]:
#separando los datos de entrenamiento y pruebas

#para los datos de entrenamiento
X_train = pd.concat([X_train_class_0, X_train_class_1, X_train_class_2, X_train_class_3, X_train_class_4, X_train_class_5, X_train_class_6]).values
y_train = pd.concat([y_train_class_0, y_train_class_1, y_train_class_2, y_train_class_3, y_train_class_4, y_train_class_5, y_train_class_6]).values


indices_train = np.random.permutation(len(X_train))
X_train = X_train[indices_train]
y_train = y_train[indices_train]
m_train = len(y_train)

#para los datos de pruebas
X_test = pd.concat([X_test_class_0, X_test_class_1, X_test_class_2, X_test_class_3, X_test_class_4, X_test_class_5, X_test_class_6]).values
y_test = pd.concat([y_test_class_0, y_test_class_1, y_test_class_2, y_test_class_3, y_test_class_4, y_test_class_5, y_test_class_6]).values

indices_test = np.random.permutation(len(X_test))
X_test = X_test[indices_test]
y_test = y_test[indices_test]
m_test = len(y_test)

Separación del 80% de datos para entrenamiento y 20% para pruebas

In [14]:
#separamos el 20% para test y el 80% para entrenamiento deld dataset
# train_dataset, test_dataset = train_test_split(dataset, test_size=0.2, random_state=42)

# Seleccionamos las columnas para X y la columna 'round_winner' para y
X_test = X_test.copy()
y_test = y_test.copy()
m_test = len(y_test)

# Seleccionamos las columnas para X y la columna 'shares' para y
X_train = X_train.copy()
y_train = y_train.copy()
m_train = len(y_train)
# Ahora, X y y deberían contener datos numéricos del 80% del dataset



# imprimir todos las X de datos
print(X_train)
print(y_train)

[[3084  112   18 ...    0    0    0]
 [2969    8    7 ...    0    0    0]
 [3289  326   19 ...    0    0    0]
 ...
 [3116   84    6 ...    0    0    0]
 [3350  328   21 ...    1    0    0]
 [3270   20    8 ...    0    0    0]]
[1 2 1 ... 2 1 1]


In [15]:
#Visualizamos una cantidad de datos de entrenamiento
print('{:10s}'.format('X[:,0]'), end='')
for i in range(1, 30):
    print('{:10.2f}'.format(X_train[0, i]), end='')
print('{:10s}'.format('Y'))

print('-' * 120)

for i in range(10):
    print('{:10.2f}'.format(X_train[i, 0]), end='')
    for j in range(1, 30):
        print('{:10.2f}'.format(X_train[i, j]), end='')
    print('{:10.2f}'.format(y_train[i]))

print(" ")
print('El 80% de ejemplos para entrenamiento son la cantidad de: {:.0f} de ejemplos'.format( len(X_train)))
print('El 20% de ejemplos para pruebas son la cantidad de: {:.0f} de ejemplos'.format( len(X_test)))
print('La cantidad total de ejemplos es de: {:.0f} de ejemplos'.format( len(dataset)))

X[:,0]        112.00     18.00    216.00    -11.00    800.00    248.00    217.00     90.00    815.00      1.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00Y         
------------------------------------------------------------------------------------------------------------------------
   3084.00    112.00     18.00    216.00    -11.00    800.00    248.00    217.00     90.00    815.00      1.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      1.00
   2969.00      8.00      7.00    150.00     19.00   2140.00    212.00    227.00    152.00   2547.00      0.00      0.00      1.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0

In [16]:
print(f"el 80% de datos es:{X_train.shape[0]} y el 20% es: {X_test.shape[0]}")
print( "la totalidad de datos es: ", dataset.shape[0])

el 80% de datos es:464807 y el 20% es: 116205
la totalidad de datos es:  581012


Imagen clara de las entradas

In [17]:
print("la cantidad de columnas es: ", X_train.shape[1])
print("la cantidad de ejemplos es: ", X_train.shape[0])

la cantidad de columnas es:  54
la cantidad de ejemplos es:  464807


Definición de las capas


*   Capa de entrada, cantidad de características con las que cuenta el dataset
*   capa oculta, se define haciendo pruebas
*   Capa de salida, cuenta con 7 clases





In [19]:
# Configurando parametros necesario
input_layer_size  = X_train.shape[1]  # Entrada de 59 caracteristicas, es la cantidad de columnas existente en el dataset
hidden_layer_size = 10   # 10 unidades ocultas en la capa oculta
num_labels = 7       # como es una clasificacion tiene solo 2 salidas

print("la cantidad de elementos en la capa de entrada es: ",input_layer_size)
print("la cantidad de elementos en la capa de oculta es: ",hidden_layer_size)
print("la cantidad de elementos en la capa de salida es: ",num_labels)

la cantidad de elementos en la capa de entrada es:  54
la cantidad de elementos en la capa de oculta es:  10
la cantidad de elementos en la capa de salida es:  7


Inicialización de Thetas

In [20]:
pesos = {}
pesos['Theta1'] = np.random.rand(hidden_layer_size , input_layer_size + 1)
pesos['Theta2'] = np.random.rand(num_labels, hidden_layer_size + 1)

In [21]:
#Se guardan los valores de theta en nuevas variables
Theta1, Theta2 = pesos['Theta1'], pesos['Theta2']

print("El tamaño de Theta1 es de: ",Theta1.ravel().shape)
print("El tamaño de Theta2 es de: ",Theta2.ravel().shape)

El tamaño de Theta1 es de:  (550,)
El tamaño de Theta2 es de:  (77,)


In [22]:
#Se forma un solo array concadenando theta1 y theta2
nn_params = np.concatenate([Theta1.ravel(), Theta2.ravel()])
print(nn_params.shape)

(627,)


**Funcion de activación**

Función Sigmoidea

In [23]:
def sigmoid(z):
    """
    Computes the sigmoid of z.
    """
    return 1.0 / (1.0 + np.exp(-z))


def sigmoidGradient(z):

    g = np.zeros(z.shape)

    g = sigmoid(z) * (1 - sigmoid(z))

    return g

In [24]:
print(nn_params.shape)

(627,)


Función de Costo

In [25]:
def nnCostFunction(nn_params,
                   input_layer_size,
                   hidden_layer_size,
                   num_labels,
                   X, y, lambda_=0.0):

    Theta1 = np.reshape(nn_params[:hidden_layer_size * (input_layer_size + 1)],
                        (hidden_layer_size, (input_layer_size + 1)))

    Theta2 = np.reshape(nn_params[(hidden_layer_size * (input_layer_size + 1)):],
                        (num_labels, (hidden_layer_size + 1)))

    m = y.size

    J = 0
    Theta1_grad = np.zeros(Theta1.shape)
    Theta2_grad = np.zeros(Theta2.shape)

    a1 = np.concatenate([np.ones((m, 1)), X], axis=1)

    a2 = sigmoid(a1.dot(Theta1.T))
    a2 = np.concatenate([np.ones((a2.shape[0], 1)), a2], axis=1)

    a3 = sigmoid(a2.dot(Theta2.T))
    y_matrix = y.reshape(-1)
    y_matrix = np.eye(num_labels)[y_matrix]

    temp1 = Theta1
    temp2 = Theta2

    reg_term = (lambda_ / (2 * m)) * (np.sum(np.square(temp1[:, 1:])) + np.sum(np.square(temp2[:, 1:])))
    J = (-1 / m) * np.sum((np.log(a3) * y_matrix) + np.log(1 - a3) * (1 - y_matrix)) + reg_term

    # Backpropogation

    delta_3 = a3 - y_matrix
    delta_2 = delta_3.dot(Theta2)[:, 1:] * sigmoidGradient(a1.dot(Theta1.T))

    Delta1 = delta_2.T.dot(a1)
    Delta2 = delta_3.T.dot(a2)

    # Agregar regularización al gradiente

    Theta1_grad = (1 / m) * Delta1
    Theta1_grad[:, 1:] = Theta1_grad[:, 1:] + (lambda_ / m) * Theta1[:, 1:]

    Theta2_grad = (1 / m) * Delta2
    Theta2_grad[:, 1:] = Theta2_grad[:, 1:] + (lambda_ / m) * Theta2[:, 1:]

    grad = np.concatenate([Theta1_grad.ravel(), Theta2_grad.ravel()])

    return J, grad

Normalización de Datos

In [26]:
def  featureNormalize(X):
    X_norm = X.copy()

    #creamos un array de ceros con una longitud igual al número de columnas en el array X. La variable mu y sigma se inicializa como este array de ceros.
    mu = np.zeros(X.shape[1])
    sigma = np.zeros(X.shape[1])

    #Creamos el promedio de cada columna de X
    mu = np.mean(X, axis = 0)
    sigma = np.std(X, axis = 0)

    sigma[sigma == 0] = 1

    #normalizamos los datos con la siguiente formula
    X_norm = (X - mu) / sigma

    return X_norm, mu, sigma

In [27]:
X_norm, mu, sigma = featureNormalize(X_train)

# imprimir todos las X_norm de datos solo 10
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>6s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]', 'X[:, 10]', 'X[:, 11]', 'X[:, 12]',
    'X[:, 13]', 'X[:, 14]', 'X[:, 15]', 'X[:, 16]', 'X[:, 17]', 'X[:, 18]', 'X[:, 19]', 'X[:, 20]', 'X[:, 21]', 'X[:, 22]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:8.0f}'.format(
        X_norm[i, 0],
        X_norm[i, 1],
        X_norm[i, 2],
        X_norm[i, 3],
        X_norm[i, 4],
        X_norm[i, 5],
        X_norm[i, 6],
        X_norm[i, 7],
        X_norm[i, 8],
        X_norm[i, 9],
        X_norm[i, 10],
        X_norm[i, 11],
        X_norm[i, 12],
        X_norm[i, 13],
        X_norm[i, 14],
        X_norm[i, 15],
        X_norm[i, 16],
        X_norm[i, 17],
        X_norm[i, 18],
        X_norm[i, 19],
        X_norm[i, 20],
        X_norm[i, 21],
        X_norm[i, 22],
        y_train[i]
    ))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]  X[:, 10]  X[:, 11]  X[:, 12]  X[:, 13]  X[:, 14]  X[:, 15]  X[:, 16]  X[:, 17]  X[:, 18]  X[:, 19]  X[:, 20]  X[:, 21]  X[:, 22]     Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      0.45     -0.39      0.52     -0.25     -0.99     -1.00      1.34     -0.32     -1.37     -0.88      1.11     -0.23     -0.88     -0.26     -0.07     -0.11     -0.09     -0.15     -0.05     -0.11     -0.01     -0.02     -0.04       1
      0.03     -1.32     -0.95     -0.56     -0.47     -0.14     -0.01      0.19      0.25      0.43     -0.90     -0.23      1.14     -0.26     -0.07     -0.11     -0.09     -0.15     -0.05     -0.11     -0.01     -0.02     -0.04       2
      1.18      1.52      0.66    

In [28]:
print("Longitud de nn_params:", len(nn_params))
print("Expected size for Theta2:", (hidden_layer_size + 1) * num_labels)

Longitud de nn_params: 627
Expected size for Theta2: 77


Cálculo del costo

In [29]:
lambda_ = 10
J, _ = nnCostFunction(nn_params, input_layer_size, hidden_layer_size, num_labels, X_norm, y_train, lambda_)
print('Costo en parametros (cargado de ex4weights): %.6f ' % J)
print('El costo debe esta cercano a               : 2.413307')

Costo en parametros (cargado de ex4weights): 17.090487 
El costo debe esta cercano a               : 2.413307


Inicialización de pesos aleatorios

In [30]:
def randInitializeWeights(L_in, L_out, epsilon_init=0.12):

    W = np.zeros((L_out, 1 + L_in))
    W = np.random.rand(L_out, 1 + L_in) * 2 * epsilon_init - epsilon_init

    return W

In [31]:
print('Inicialización de parámetros de redes neuronales...')

initial_Theta1 = randInitializeWeights(input_layer_size, hidden_layer_size)
initial_Theta2 = randInitializeWeights(hidden_layer_size, num_labels)

# Desenrrollr parametros
initial_nn_params = np.concatenate([initial_Theta1.ravel(), initial_Theta2.ravel()], axis=0)

Inicialización de parámetros de redes neuronales...


Cálculo de Thetas

In [32]:
# Después de haber completado la tarea, cambie el maxiter a uno más grande
# Usando un numero maximo de iteraciones
options= {'maxiter': 1000}

# Tratamos con diferentes valores de lambda_
lambda_ = 1

# define una función anónima (lambda) llamada costFunction.
costFunction = lambda p: nnCostFunction(p, input_layer_size,
                                        hidden_layer_size,
                                        num_labels, X_norm, y_train, lambda_)

# Ahora, costFunction es una función que toma solo un argumento.
# (los parámetros de la red neuronal)
res = optimize.minimize(costFunction,
                        initial_nn_params,
                        jac=True,
                        method='TNC',
                        options=options)

# obtenemos la solución de la optimización
nn_params = res.x

# Obtenemos Theta1 y Theta2 de nn_params
Theta1 = np.reshape(nn_params[:hidden_layer_size * (input_layer_size + 1)],
                    (hidden_layer_size, (input_layer_size + 1)))

Theta2 = np.reshape(nn_params[(hidden_layer_size * (input_layer_size + 1)):],
                    (num_labels, (hidden_layer_size + 1)))

<ipython-input-32-fd6245226a93>:15: OptimizeWarning: Unknown solver options: maxiter
  res = optimize.minimize(costFunction,
<ipython-input-23-fcf1f7fde265>:5: RuntimeWarning: overflow encountered in exp
  return 1.0 / (1.0 + np.exp(-z))


In [33]:
for i in range(len(Theta1)):
    print("Thetha",i,"es: ", Theta1[i])
    print("")

Thetha 0 es:  [ 3.4670124   0.0318579  -0.04171384 -0.53216769 -0.25587558  0.06122593
 -0.30165831  0.18559381 -0.4909917   0.67987679 -1.98508737  4.08669125
  1.26725351 -2.39422027 -4.51357419  0.00524558  0.28997098 -0.57557336
 -0.00591526  0.08531776  0.0738383  -0.14313    -0.11664456 -0.5382063
  0.05090093  1.2928416   0.90171689  0.28742578 -0.30399993 -0.03959287
 -1.4481487  -0.59185256 -0.19749     0.09541037  1.30755952 -0.28672571
 -2.63109727  0.33617119  0.75681801 -0.1283539   0.78027193  0.1990728
  0.04868328 -0.75746961  0.72694659  0.23479298  1.37018297 -0.10833304
 -0.02711405 -0.19988199  0.01886436 -0.38388645  0.5483308  -2.16394244
 -0.94602312]

Thetha 1 es:  [ 5.06798925e+00  1.12342289e-01 -1.25638980e-01  7.16442107e-02
 -2.63512402e-01 -9.98581126e-02  6.53221815e-01  1.67200929e+00
 -1.44220877e+00  2.57000130e+00  7.39769037e-01  8.77347855e-02
  2.41979670e+00 -6.70970071e-01 -7.35173289e-01 -6.52818646e-02
 -3.55234188e-01 -4.58229346e-01 -5.155514

In [34]:
for i in range(len(Theta2)):
    print("Thetha",i,"es: ", Theta1[i])
    print("")

Thetha 0 es:  [ 3.4670124   0.0318579  -0.04171384 -0.53216769 -0.25587558  0.06122593
 -0.30165831  0.18559381 -0.4909917   0.67987679 -1.98508737  4.08669125
  1.26725351 -2.39422027 -4.51357419  0.00524558  0.28997098 -0.57557336
 -0.00591526  0.08531776  0.0738383  -0.14313    -0.11664456 -0.5382063
  0.05090093  1.2928416   0.90171689  0.28742578 -0.30399993 -0.03959287
 -1.4481487  -0.59185256 -0.19749     0.09541037  1.30755952 -0.28672571
 -2.63109727  0.33617119  0.75681801 -0.1283539   0.78027193  0.1990728
  0.04868328 -0.75746961  0.72694659  0.23479298  1.37018297 -0.10833304
 -0.02711405 -0.19988199  0.01886436 -0.38388645  0.5483308  -2.16394244
 -0.94602312]

Thetha 1 es:  [ 5.06798925e+00  1.12342289e-01 -1.25638980e-01  7.16442107e-02
 -2.63512402e-01 -9.98581126e-02  6.53221815e-01  1.67200929e+00
 -1.44220877e+00  2.57000130e+00  7.39769037e-01  8.77347855e-02
  2.41979670e+00 -6.70970071e-01 -7.35173289e-01 -6.52818646e-02
 -3.55234188e-01 -4.58229346e-01 -5.155514

Función para las predicciones

In [35]:
def predict(Theta1, Theta2, X):
    # Useful values
    m = X.shape[0]
    num_labels = Theta2.shape[0]

    # You need to return the following variables correctly
    p = np.zeros(m)
    h1 = sigmoid(np.dot(np.concatenate([np.ones((m, 1)), X], axis=1), Theta1.T))
    h2 = sigmoid(np.dot(np.concatenate([np.ones((m, 1)), h1], axis=1), Theta2.T))
    p = np.argmax(h2, axis=1)
    return p

In [37]:
#Creamos la matriz con 6 ejemplos para hacer la prueba predecir un precio de cada ejemplo:
matriz_datos = np.array([
    [2596,51,3,258,0,510,221,232,148,6279,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0],
    [2590,56,2,212,-6,390,220,235,151,6225,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0],
    [2880,209,17,216,30,4986,206,253,179,4323,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0],
    [2722,315,24,30,19,3216,148,212,200,6132,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [2860,31,10,295,98,3644,218,218,135,5904,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [2909,57,16,134,23,5502,229,203,102,4084,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0]
])


#Normalizamos todos
matriz_datos= (matriz_datos - mu) / sigma


#Calculamos la Y predicha de los 6 ejemplos de prediccion
pred = predict(Theta1, Theta2, matriz_datos)

print("las predicciones para los 6 ejemplos son:")
i = 0
for i in range(len(matriz_datos)):
    print("Para el ejemplo ",i+1," es:", pred[i])
    print("")


las predicciones para los 6 ejemplos son:
Para el ejemplo  1  es: 2

Para el ejemplo  2  es: 2

Para el ejemplo  3  es: 2

Para el ejemplo  4  es: 2

Para el ejemplo  5  es: 2

Para el ejemplo  6  es: 2



Validaciones

In [38]:
X_test = (X_test - mu)/ sigma

pred = predict(Theta1, Theta2, X_test)
print(pred)

for i in range(30):
    print(f"Prediccion: {pred[i]}, Real: {y_test[i]}")

[2 2 2 ... 1 1 2]
Prediccion: 2, Real: 2
Prediccion: 2, Real: 2
Prediccion: 2, Real: 2
Prediccion: 2, Real: 1
Prediccion: 2, Real: 2
Prediccion: 2, Real: 1
Prediccion: 2, Real: 2
Prediccion: 0, Real: 0
Prediccion: 2, Real: 2
Prediccion: 2, Real: 2
Prediccion: 0, Real: 1
Prediccion: 2, Real: 2
Prediccion: 2, Real: 2
Prediccion: 1, Real: 1
Prediccion: 6, Real: 3
Prediccion: 2, Real: 2
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 3, Real: 6
Prediccion: 2, Real: 2
Prediccion: 3, Real: 3
Prediccion: 2, Real: 2
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 0
Prediccion: 3, Real: 3
Prediccion: 2, Real: 2
Prediccion: 6, Real: 6
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1


In [39]:
print('Precisión de entrenamiento: {:.2f} %'.format(np.mean(pred == y_test) * 100))

Precisión de entrenamiento: 77.02 %


Se puede determinar que el modelo reliza las predicciones casi exactas,dándonos una prediccion de 77.02 %. Tener el modelo con esta precisión puede llegar a generar desconfianza.